lib/internal/process/promises.js

```diff
@@ -18,32 +20,51 @@ function setupPromises(scheduleMicrotasks) {
 
   function unhandledRejection(promise, reason) {
     hasBeenNotifiedProperty.set(promise, false);
+    promiseToGuidProperty.set(promise, lastPromiseId++);
     addPendingUnhandledRejection(promise, reason);
   }
 
   function rejectionHandled(promise) {
     var hasBeenNotified = hasBeenNotifiedProperty.get(promise);
     if (hasBeenNotified !== undefined) {
       hasBeenNotifiedProperty.delete(promise);
+      const uid = promiseToGuidProperty.get(promise);
+      promiseToGuidProperty.delete(promise);
       if (hasBeenNotified === true) {
         process.nextTick(function() {
-          process.emit('rejectionHandled', promise);
+          if (!process.emit('rejectionHandled', promise)) {
+            process.emitWarning(
+              'Promise rejection was handled asynchronously ' +
+              `(rejection id: ${uid}) `,
+              'PromiseRejectionHandledWarning');
+          }
         });
       }
 
     }
   }
 
+  var deprecationWarned = false;
   function emitPendingUnhandledRejections() {
-    var hadListeners = false;
+    let hadListeners = false;
     while (pendingUnhandledRejections.length > 0) {
-      var promise = pendingUnhandledRejections.shift();
-      var reason = pendingUnhandledRejections.shift();
+      const promise = pendingUnhandledRejections.shift();
+      const reason = pendingUnhandledRejections.shift();
       if (hasBeenNotifiedProperty.get(promise) === false) {
         hasBeenNotifiedProperty.set(promise, true);
+        const uid = promiseToGuidProperty.get(promise);
         if (!process.emit('unhandledRejection', reason, promise)) {
-          // Nobody is listening.
-          // TODO(petkaantonov) Take some default action, see #830
+          process.emitWarning('Unhandled promise rejection ' +
+                              `(rejection id: ${uid}): ${reason}`,
+                              'UnhandledPromiseRejectionWarning');
+          if (!deprecationWarned) {
+            deprecationWarned = true;
+            process.emitWarning(
+              'Unhandled promise rejections are deprecated. In the future, ' +
+              'an error will be thrown when a promise rejection is not ' +
```

Comment:
No error will be thrown.

There is no way to catch it so that you cannot listen to GC.
It will exit with the rejection formatted how we format regular uncaught exceptions.

`'In the future, the Node process will exit if promise rejections are not handled before garbage collection.'`

That may be slightly confusing since this warning message is not on GC... not sure how to clarify.